fineract-provider/src/main/java/org/apache/fineract/cob/service/LoanCOBCatchUpServiceImpl.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.fineract.infrastructure.core.domain.FineractContext;
+import org.apache.fineract.infrastructure.core.serialization.GoogleGsonSerializerHelper;
+import org.apache.fineract.infrastructure.core.service.ThreadLocalContextUtil;
+import org.apache.fineract.infrastructure.jobs.data.JobParameterDTO;
+import org.apache.fineract.infrastructure.jobs.domain.ScheduledJobDetail;
+import org.apache.fineract.infrastructure.jobs.domain.ScheduledJobDetailRepository;
+import org.apache.fineract.infrastructure.jobs.exception.JobNotFoundException;
+import org.apache.fineract.infrastructure.jobs.service.JobStarter;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepository;
+import org.springframework.batch.core.Job;
+import org.springframework.batch.core.JobParametersInvalidException;
+import org.springframework.batch.core.configuration.JobLocator;
+import org.springframework.batch.core.launch.NoSuchJobException;
+import org.springframework.batch.core.repository.JobExecutionAlreadyRunningException;
+import org.springframework.batch.core.repository.JobInstanceAlreadyCompleteException;
+import org.springframework.batch.core.repository.JobRestartException;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class LoanCOBCatchUpServiceImpl implements LoanCOBCatchUpService, InitializingBean {
+
+    private final LoanRepository loanRepository;
+    private final JobStarter jobStarter;
+    private final JobLocator jobLocator;
+    private final ScheduledJobDetailRepository scheduledJobDetailRepository;
+    private Gson gson;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        this.gson = GoogleGsonSerializerHelper.createSimpleGson();
+    }
+
+    @Override
+    public OldestCOBProcessedLoanDTO getOldestCOBProcessedLoan() {
+        List<Loan> oldestCOBProcessedLoan = loanRepository.findOldestCOBProcessedLoan();
+        OldestCOBProcessedLoanDTO oldestCOBProcessedLoanDTO = new OldestCOBProcessedLoanDTO();
+        oldestCOBProcessedLoanDTO.setLoanIds(oldestCOBProcessedLoan.stream().map(Loan::getId).toList());
+        oldestCOBProcessedLoanDTO
+                .setCobProcessedDate(oldestCOBProcessedLoan.stream().map(Loan::getLastClosedBusinessDate).findFirst().get());
+        oldestCOBProcessedLoanDTO.setCobBusinessDate(ThreadLocalContextUtil.getBusinessDateByType(BusinessDateType.COB_DATE));
+        return oldestCOBProcessedLoanDTO;
+    }
+
+    @Override
+    public void executeLoanCOBCatchUp() {
+        FineractContext context = ThreadLocalContextUtil.getContext();
+        CompletableFuture<Void> executeCatchUpFuture = CompletableFuture.runAsync(() -> executeLoanCOBCatchUpAsync(context));
+    }
+
+    private void executeLoanCOBCatchUpAsync(FineractContext context) {
+        ThreadLocalContextUtil.init(context);
```

Comment:
Since this is being executed on a separate thread, let's put it into a try-finally block to make sure we clean up the ThreadLocalContextUtil (with a reset call).